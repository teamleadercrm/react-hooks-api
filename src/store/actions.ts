import { createAction, ActionType } from 'typesafe-actions';

import { Response } from '../typings/API';
import { CACHE_QUERY_RESULT } from './constants';

// we only have one action right now
export type RootAction = ActionType<typeof cacheQueryResult>;

type Payload = {
  domain: string;
  action: string;
  options: object;
  data: Response;
};

export const cacheQueryResult = createAction(CACHE_QUERY_RESULT)<Payload>();
