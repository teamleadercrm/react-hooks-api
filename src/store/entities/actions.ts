import { createAction } from 'typesafe-actions';

import { SAVE_NORMALIZED_ENTITIES } from './constants';
import { NormalizedEntities } from './entities';

export const saveNormalizedEntities = createAction(SAVE_NORMALIZED_ENTITIES)<{
  type: string;
  entities: NormalizedEntities;
}>();
