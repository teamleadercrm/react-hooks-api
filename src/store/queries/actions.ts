import { createStandardAction } from 'typesafe-actions';

import { QUERY_REQUEST, QUERY_FAILURE, QUERY_SUCCESS } from './constants';

export const queryRequest = createStandardAction(QUERY_REQUEST)<{ key: string }>();
export const queryFailure = createStandardAction(QUERY_FAILURE)<{ key: string; error: Error }>();
export const querySuccess = createStandardAction(QUERY_SUCCESS)<{ key: string; ids?: string | string[]; data?: any, meta?: any }>();
