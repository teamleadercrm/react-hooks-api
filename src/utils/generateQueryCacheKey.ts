// @TODO replace this with fast-stringify
const generateQueryCacheKey = (obj: object) => JSON.stringify(obj);

export default generateQueryCacheKey;
