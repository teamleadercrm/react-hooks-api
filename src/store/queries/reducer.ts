import { DeepReadonly } from 'utility-types';
import { ActionType, getType } from 'typesafe-actions';
import produce, { Draft } from 'immer';

import * as actions from './actions';

export type State = DeepReadonly<{
  [key: string]: {
    loading: boolean;
    error?: Error;
    ids?: string | string[];
    data?: any,
    meta?: any;
  };
}>;

export type QueryAction = ActionType<typeof actions>;

const INITIAL_STATE: State = {};

const queries = produce((draft: Draft<State>, action: QueryAction) => {
  switch (action.type) {
    case getType(actions.queryRequest):
      draft[action.payload.key] = { ...draft[action.payload.key], loading: true };
      return;
    case getType(actions.queryFailure):
      draft[action.payload.key] = { ...draft[action.payload.key], loading: false, error: action.payload.error };
      return;
    case getType(actions.querySuccess):
      draft[action.payload.key] = {
        ...draft[action.payload.key],
        loading: false,
        error: undefined,
        ids: action.payload.ids,
        data: action.payload.data,
        meta: action.payload.meta,
      };
      return;
  }
}, INITIAL_STATE);

export default queries;
