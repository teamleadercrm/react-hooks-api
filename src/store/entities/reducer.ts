import { DeepReadonly } from 'utility-types';
import { ActionType, getType } from 'typesafe-actions';
import produce, { Draft } from 'immer';

import * as actions from './actions';
import { NormalizedEntities } from './entities';

export type State = DeepReadonly<{
  [key: string]: NormalizedEntities;
}>;

export type EntitiesAction = ActionType<typeof actions>;

const INITIAL_STATE: State = {};

const entities = produce((draft: Draft<State>, action: EntitiesAction) => {
  switch (action.type) {
    case getType(actions.saveNormalizedEntities):
      draft[action.payload.type] = Object.keys(action.payload.entities).reduce(
        (entitiesById, nextEntityId) => ({
          ...entitiesById,
          [nextEntityId]: {
            ...entitiesById[nextEntityId],
            ...action.payload.entities[nextEntityId],
          },
        }),
        draft[action.payload.type] || {}
      );
      return;
  }
}, INITIAL_STATE);

export default entities;
