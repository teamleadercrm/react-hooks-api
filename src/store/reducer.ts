import { DeepReadonly } from 'utility-types';
import { createReducer } from 'typesafe-actions';

import { CACHE_QUERY_RESULT } from './constants';
import generateQueryCacheKey from '../utils/generateQueryCacheKey';
import { Response } from '../typings/API';

import { RootAction } from './actions';

export type State = DeepReadonly<{
  queries: {
    [key: string]: Response;
  };
}>;

const INITIAL_STATE: State = {
  queries: {},
};

const queries = createReducer<State, RootAction>(INITIAL_STATE).handleAction(
  CACHE_QUERY_RESULT,
  (state, { payload }) => {
    const key = generateQueryCacheKey({ domain: payload.domain, action: payload.action, options: payload.options });
    return { ...state, queries: { ...state.queries, [key]: payload.data } };
  },
);

export default queries;
