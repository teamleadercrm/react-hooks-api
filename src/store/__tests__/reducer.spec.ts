import reducer, { State } from '../reducer';
import { cacheQueryResult } from '../actions';
import generateQueryCacheKey from '../../utils/generateQueryCacheKey';

const getInitialState = (initial?: Partial<State>) => reducer(initial as State, {} as any);

describe('reducer', () => {
  it('initial', () => {
    const INITIAL_STATE = getInitialState();
    expect(INITIAL_STATE).toMatchSnapshot();
  });

  it('CACHE_QUERY_RESULT', () => {
    const INITIAL_STATE = getInitialState();

    const payload = {
      domain: 'users',
      action: 'list',
      options: {},
      data: { data: {} },
    };

    const key = generateQueryCacheKey({
      domain: payload.domain,
      action: payload.action,
      options: payload.options,
    });

    const resultState = {
      queries: {
        [key]: payload.data,
      },
    };

    expect(reducer(INITIAL_STATE, cacheQueryResult(payload))).toEqual(resultState);
  });
});
