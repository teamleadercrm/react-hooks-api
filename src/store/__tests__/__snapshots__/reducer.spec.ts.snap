// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer initial 1`] = `
Object {
  "queries": Object {},
}
`;
